### ISSUE TITLE
Icons take several seconds to appear/disappear from the dock when the app is opened/closed

### ISSUE THREAD

**Issue by JohnLBergqvist** [micheleg/dash-to-dock#1173]
When opening and closing an application, there is a delay of up to 30 seconds before that icon appears or disappears from the dock. I'm running Ubuntu 20.04 beta with Dash to Dock version 68. The same problem occured on both this one, and Ubuntu's own forked version that was installed by default, and it occurs when running both free (Nouveau) & proprietory Nvidia drivers.

**Comment by thiagola92** [comment]
The icons doesn't disappear at all to me

**Comment by sonOfRa** [comment]
I'm having the same issue since upgrading to Focal Fossa on friday. Doesn't matter if I use the builtin "Ubuntu Dock" version, or the manually installed Dash-To-Dock. Haven't tried nouveau drivers, I'm on the proprietary nvidia drivers.

**Comment by JohnLBergqvist** [comment]
@thiagola92 Yes, a couple of days before the full release (I was on the Foscal beta, now on the full release), the issue resolved itself via a software update (although i'm not sure which package fixed the problem, it certainly wasn't an update to the extension itself).

**Comment by felipehw** [comment]
I'm suffering with this problem at Fedora 32 + Wayland + Dash To Dock v68.

**Comment by rootinshik** [comment]
I have the same problem too

**Comment by rootinshik** [comment]
https://askubuntu.com/questions/1234400/icons-take-several-seconds-to-appear-disappear-from-the-dock-when-the-app-is-ope

**Comment by Gonm1** [comment]
I have the same issue on Pop!_OS 20.04

**Comment by feedspeed** [comment]
+1 with same issue, latest Dash to Dock (v68) with Fedora 32 + Wayland

**Comment by Gonm1** [comment]
I have installed the Dash to Panel extension and the problem persists. Went ahead and uninstalled both extensions and the problem is also present in the default panel over the activities overview. Any information on this please share.
I have GNOME 3.36.1

**Comment by getaceres** [comment]
Same here in a freshly installed Ubuntu 20.04. When booting it takes some seconds to show the icons on the dock and when adding or removing applications to the dock it takes some time to react too. I click on "Remove from favorites" and then it stays in the dock for some seconds and then it dissappears.

**Comment by felipehw** [comment]
> I have installed the Dash to Panel extension and the problem persists. Went ahead and uninstalled both extensions and the problem is also present in the default panel over the activities overview. Any information on this please share.
> I have GNOME 3.36.1

I solved this problem by disabling this extension and restarting the machine.
Just disable the extension without a restart ... don't resolve my problem. The problem affects the vanilla dash until I restarted the machine.

By now ... I'll stop my use of this extension.

**Comment by kulomb** [comment]
Icons don't refresh at all on Ubuntu 20.04 with GNOME 3.36.2 and Dash To Dock 68.

**Comment by Ian-FR** [comment]
+1 with this issue, latest Dash to Dock (v69) with Manjaro 21.1.3 (Pahvo) + Wayland

**Comment by NathanRussell22** [comment]
I'm having the same issue on `Manjaro 21.2.2 (Gnome 41)` With Dash to Dock Version `71` in both `Xorg` & `Wayland`.
I have tried re-installing the extension and building it manually but I cannot seem to fix this issue.

**Comment by Alt37** [comment]
I can easily reproduce this bug on Ubuntu 22.04 (Xorg session)
1) Create simple CSV file
`printf '"entity_id","type_id"' > libreoffice.csv`
2) Open it with Libreoffice Calc
3) On "Text Import" dialog press "OK"
4) "Text import" window closes, new LibreOffice window appears
5) Icon is missing until you enter "Dash to Dock Settings", reload shell etc
I wasn't able to reproduce this case with Wayland and original Dash

[dash-to-dock-x11.webm](https://user-images.githubusercontent.com/44649402/199135689-61a8ce0c-024e-44cc-ab97-4c373bb3a40e.webm)

[dash-to-dock-wayland.webm](https://user-images.githubusercontent.com/44649402/199136105-ec822d7b-61f9-447f-80aa-f8c6fd6f5b82.webm)

[original-dash.webm](https://user-images.githubusercontent.com/44649402/199136112-dab94d9e-bd19-406b-b218-5007a2e230e6.webm)

**Comment by MathisP75** [comment]
I am having the same problem with some apps on Pop OS 22.04 LTS, on both Wayland and Xorg.

**Comment by roboknight** [comment]
Can't be sure if this is the exact same problem, but I have a multi-monitor setup.  When I run a java application from a bash script, the application shows up, but the dock icon is blank.  Checking the application with xprops shows what I would expect for WM_CLASS, as well as several fixed sized bitmaps for the application's icon.  The dock icon appears if I move the application window from one monitor to another.  There are instances where the application's icon appears upon running, but usually it doesn't appear.  I don't know what triggers the application icon to appear in the dock in those cases, but it always re-appears when the window is moved to a different monitor.  This is Ubuntu 20.04 which appears to have the ubuntu-dock extension installed (which based on what I read about that extension seems to be a placeholder for dash to dock).  I can't tell what event occurs that makes it suddenly display the icon in the dock, but the icon doesn't even appear after "long time" if I just let the application sit there.  It seems to be, if not the same problem, closely related.  Maybe it will help someone track this down.  I've tried all kinds of things to get the icon to appear when run initially without any luck.  Sometimes it just does.  Not sure why.

**Comment by Alt37** [comment]
@roboknight 
> The application does open a splash screen then, if no project is opened, a window remains opened, so maybe something is just getting fouled up with which window is being latched onto for display of the icon.
Looks like it really has something to do with the "first" window (splash screen, dialog etc) and "main" window. From my test above with LibreOffice CSV import the "first window" is a dialog (_NET_WM_WINDOW_TYPE_DIALOG), which is replaced by the "main" window after import.
The only good workaround I've found for now is to "add to favourites" application "pinning" it to the panel.

**Comment by roboknight** [comment]
As another point, if I run the application, move the window to a new monitor (producing an icon), then open the application AGAIN, even though the WM_CLASS is the same, the second instance never displays the icon, even if I move it to a new Window.  So, the second instance NEVER gets an icon (and also doesn't "nest" in the first icon even though the WM_CLASS is the same).  Since this application that I'm running through a bash script is Java based, it uses Swing.  I don't know what effect this has (and is the reason I wasn't ENTIRELY sure this was the same problem), but as I noted above, the second instance NEVER gets an icon, even if I move its main window to a new monitor as I do the first instance.  I've tried to figure out what even is even causing the monitor switch to "clear up" the issue the first time, but when it comes to tracking down GUI issues, I'm not sure what tools would really help here.  Others more familiar with GUI development maybe could clue me in so I could poke around and maybe figure out what is really going on?

**Comment by Alt37** [comment]
Also reproducible with the final dialog of "usb-creator-gtk"
Both "main window" and "Installation finished" dialog share the same WM_CLASS
`WM_CLASS(STRING) = "usb-creator-gtk", "Usb-creator-gtk"`
but obviously got different "WINDOW_TYPE"
"Main window"
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_NORMAL
"Installation finished" dialog
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_DIALOG
